README.md
### NFT art maker

The primary task of this tool is to generate a randomized set of images from provided PNG layers and pack them into .car files. Then you can upload them to IPFS using [nft.storage](https://nft.storage/) or other tools.

**Please test it before using it for the real stuff. It can always be buggy.**

- PNG layers for testing: https://ipfs.io/ipfs/bafkreicwfzj7f3xc6mjkyaqknd4gsosscznelpiwdtwmdp773irwuv2lqu (or check the [example](https://github.com/juliancwirko/nft-art-maker/tree/main/example) directory)
- Latest walkthrough video: https://youtu.be/S26Zskop-aU
- Older video, as part of bigger process on Elrond blockchain: https://youtu.be/resGP6a7_34

#### For older versions (check changelog):
- [CHANGELOG](https://github.com/juliancwirko/nft-art-maker/blob/main/CHANGELOG.md)

#### Based on:
- [HashLips art engine](https://github.com/HashLips/hashlips_art_engine) - only main functionality (output metadata schema should be suitable for most of the marketplaces)
- [Pixels to SVG](https://codepen.io/shshaw/pen/XbxvNj) - SVG code from images
- [SVGO](https://github.com/svg/svgo) - SVG optimization and base64 data uri generation
- [ipfs-car](https://github.com/web3-storage/ipfs-car) - optionally for handling ipfs .car archives
- [dotProp](https://github.com/sindresorhus/dot-prop) - for better managing the metadata schema mapper

The random assets generation is a customized and simplified version of the [HashLips art engine](https://github.com/HashLips/hashlips_art_engine) logic. If you need more options and functionality, please use HashLips.

**The tool offers two different outputs:**
1. png and metadata files packed into the ipfs .car files. Base image CID will be updated in all metadata files automatically after running `nft-art-maker pack` and base CID for metadata files will be added to the summary metadata json file.
2. (experimental) all data with encoded svgs in one big metadata.json file, without any additional files. This will be useful when you want to have non-standard on-chain only nfts. Be aware that the SVG output can be buggy on very complicated and big images. This experimental option is for small simple images, like pixel art etc.

nft-art-maker tool doesn't assume any way of uploading to ipfs, but I would recommend [nft.storage](https://nft.storage/) where you can upload whole .car file. They offer free pinning service and Filecoin storage. So even if you delete it there or nft.storage stops working for some reason, the data will persist. Of course, learn about it first. They have a friendly UI, but you can also use the CLI tool for that.

### Table of Contents  
- [How to use it](#how-to-use-it)  
- [Configuration options](#configuration-options)
- [Layers configuration](#layers-configuration)
- [Shuffle layers](#shufflelayerconfigurations)
- [Output types](#output-types-configuration)
- [Output metadata schema configuration](#output-metadata-schema-configuration)
- [Layers directory structure (example)](#layers-directory-structure-example)
- [Upload packed car files to nft.storage](#upload-packed-car-files-to-nftstorage)
- [Update images base CID](#update-images-base-cid)
- [Try the example](#try-the-example)

#### How to use it:
- minimum version of Node is **14.14.0**. Best **18**.
- create a project directory -> `mkdir my-nft-collection ; cd my-nft-collection`
- in that directory, create the `layers` directory with all of your layers split into proper directories. Read more about it below.
- create a configuration file `.nftartmakerrc` (other file names also allowed, check out [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) for more info). This file should be a JSON formatted config file. You'll find all configuration options below.
- run `npx nft-art-maker generate` or if installed globally `nft-art-maker generate` - it will generate all files or encoded SVG with metadata json file for each, plus it will generate one big metadata file with all editions and provenance hash. Additionally, there is also an option to pack all images and metadata files into an ipfs car files. Then every CIDs will be also actual in metadata files.

You can always install it globally by `npm install nft-art-maker -g` and then use it like `nft-art-maker generate`.

Updating: when using npx, make sure that it takes the new version. You can always install it globally using `npm install nft-art-maker@latest -g`.

#### Additionally you can:
- generate a preview - run `npx nft-art-maker preview`
- you can also pack files using `npx nft-art-maker pack` (always recommended!) - this will pack all files using ipfs-car into one images.car and metadata.car files, which you can upload using services like nft.storage
- check how many unique assets is generated. `npx nft-art-maker check` Sometimes the names of files can be misleading when there are not enough layers to achieve the required amount of assets. This probably needs some rewrites in the future.
- upload packed car files to nft.storage. `npx nft-art-maker upload` - this will upload all packed ipfs-car files to nft.storage. Set "nftStorageApiToken" config option to your API token
- overwrite image CID. `npx nft-art-maker updateImageCID` - this will update all metadata files based on the config option `overwriteImageCID`

#### Configuration options

You should use the config file at least for layers configuration. But there are also other configuration options. Whole config example:

```json
{
  "description": "Your collection name",
  "svgBase64DataOnly": true,
  "layerConfigurations": [
    {
      "growEditionSizeTo": 100,
      "layersOrder": [
        {
          "name": "face"
        },
        {
          "name": "head"
        },
        {
          "name": "eyes"
        }
      ]
    },
    {
      "growEditionSizeTo": 110,
      "layersOrder": [
        {
          "name": "pinkyFace"
        },
        {
          "name": "head"
        },
        {
          "name": "eyes"
        }
      ]
    }
  ],
  "shuffleLayerConfigurations": false,
  "format": {
    "width": 20,
    "height": 20
  },
  "metadataSchemaMapper": {
    "name": "name",
    "description": "description",
    "edition": "edition",
    "attributes": "attributes",
    "base64SvgDataUri": "base64SvgDataUri",
    "image.href": "image.href",
    "image.hash": "image.hash",
    "image.ipfsUri": "image.ipfsUri",
    "image.ipfsCid": "image.ipfsCid",
    "image.fileName": "image.fileName"
  },
  "rarityDelimiter": "#",
  "uniqueDnaTorrance": 10000,
  "layersDirName": "layers",
  "outputDirName": "output",
  "outputJsonDirName": "json",
  "outputImagesDirName": "images",
  "outputJsonFileName": "metadata.json",
  "outputImagesCarFileName": "images.car",
  "outputMetadataCarFileName": "metadata.car",
  "outputMetadataFileExtension": ".json",
  "editionNameFormat": "#",
  "tags": "tag1,tag2,tag3",
  "preview": {
    "thumbPerRow": 20,
    "thumbWidth": 60,
    "imageRatio": 1,
    "imageName": "preview.png"
  },
  "nftStorageApiToken": "Your nft.storage API token",
  "overwriteImageCID": "Your image CID after upload to your IPFS storage."
}
```

##### layers configuration

Every subdirectory in your `layers` directory should be named after the type of the layer, and inside, you should put your png files. The name structure should be as follows: `filename1#100.png`, where the `#100` is a rarity weight. Good explainer in Hashlips repo: [here](https://github.com/HashLips/hashlips_art_engine/issues/251#issuecomment-969911889).

##### shuffleLayerConfigurations

Enabled by default, but you can always enable it to shuffle items from different `layerConfigurations`.

##### Output types configuration

You can decide if you want to have encoded SVGs or standard PNGs files. Use `svgBase64DataOnly` setting. **Be aware** that these are totally separate use cases, not the option to choose the file format.

The example of output `metadata.json` file structure with empty values (this is one big file with all editions):

```json
{
  "editions": [
    {
      "name": "",
      "description": "",
      "edition": 0,
      "attributes": [
        {
          "trait_type": "",
          "value": ""
        }
      ],
      "base64SvgDataUri": "",
      "image": {
        "href": "",
        "hash": "",
        "ipfsUri": "",
        "ipfsCid": "",
        "fileName": ""
      }
    }
  ],
  "provenanceHash": "",
  "metadataFilesIpfsBaseCid": ""
}

```

The example of a single output metadata file (for example 1.json):

```json
{
  "name": "",
  "description": "",
  "edition": 0,
  "attributes": [
    {
      "trait_type": "",
      "value": ""
    }
  ],
  "base64SvgDataUri": "",
  "image": {
    "href": "",
    "hash": "",
    "ipfsUri": "",
    "ipfsCid": "",
    "fileName": ""
  }
}
```

##### Output metadata schema configuration

From v5.2.0, you can modify the structure of the final metadata files. For now, if you would like to do that, you should configure all fields. Here is the default configuration (check how to use it in the .nftartmakerrc config file above):

```json
{
  "metadataSchemaMapper": {
    "name": "name",
    "description": "description",
    "edition": "edition",
    "attributes": "attributes",
    "base64SvgDataUri": "base64SvgDataUri",
    "image.href": "image.href",
    "image.hash": "image.hash",
    "image.ipfsUri": "image.ipfsUri",
    "image.ipfsCid": "image.ipfsCid",
    "image.fileName": "image.fileName"
  }
}
```

As you can see here, we have a `key:value` object. All keys are currently required variables, which should all be in the configuration if you decide to use the `metadataSchemaMapper`.

So, for example, let's change the structure to match the OpenSea requirements. Because not all fields are required by OpenSea, we will put them in a separate key.

OpenSea metadata mapper proposition example:
```json
{
  "metadataSchemaMapper": {
    "name": "name",
    "description": "description",
    "edition": "edition",
    "attributes": "attributes",
    "base64SvgDataUri": "base64SvgDataUri",
    "image.href": "external_url",
    "image.hash": "image_additional_info.hash",
    "image.ipfsUri": "image",
    "image.ipfsCid": "image_additional_info.ipfsCid",
    "image.fileName": "image_additional_info.fileName"
  }
}
```

We changed the `image.href` to be `external_url` in our metadata files and `image.ipfsUri` to `image` in our metadata files. We also moved all other data from the default `image` key to the separate `image_additional_info` key. So now our final structure looks like this: 

OpenSea final metadata form example:
```json
{
  "name": "",
  "description": "",
  "edition": 1,
  "attributes": [
    {
      "trait_type": "",
      "value": ""
    }
  ],
  "external_url": "",
  "image_additional_info": {
    "hash": "",
    "ipfsCid": "",
    "fileName": ""
  },
  "image": ""
}
```

Deployed example: https://bafybeih6ii5v2hmeqsiwzkvqqw7oearegyysr6ibgpzfgmlf3ancjtghku.ipfs.nftstorage.link/111.json

**Important!**
Some of the `image` fields will be populated appropriately after running the `pack` command. All JSON files will get an update. They will be also updated in the final `metadata.car` file. This is why it is always recommended to use the `pack` command. But you can also update all paths with your own scripts.

**The example of an actual single metadata file**: https://ipfs.io/ipfs/bafybeied5z3gndksdmbeaqxgbxryax6xp27nqfj3wy7c5x3usrol4sngam/149.json (the image is just for tests, poor quality, but also generated using the tool, there is a whole collection of 200 pieces so that you can change numbers to preview them).

##### Layers directory structure (example)

Here is an example of the `layers` directory structure with the configuration from the example above.

```bash
layers/
├── eyes
│   ├── eyes1#100.png
│   ├── eyes2#100.png
│   ├── eyes3#100.png
│   ├── eyes4#100.png
│   ├── eyes5#100.png
│   └── eyes6#100.png
├── face
│   ├── face1#100.png
│   ├── face2#100.png
│   ├── face3#100.png
│   ├── face4#100.png
│   └── face5#100.png
└── head
    ├── head1#100.png
    ├── head2#100.png
    ├── head3#100.png
    ├── head4#100.png
    ├── head5#100.png
    ├── head6#100.png
    ├── head7#100.png
    └── head8#100.png
```

### Upload packed car files to nft.storage

This package has a small helper to upload the packed car files to nft.storage. To archive the only requirement is addition in the .nftartmakerrc config file: `"nftStorageApiToken": "Your nft.storage API token"`. See the example above.

### Update images base CID

**Important!** It is not a part of the standard flow! `pack` command will do this for you too.

This is useful when you want to upload image folder with NFT UP or Pinata (or any other 3rd party IPFS storage provider.)

In case you are dealing with a very large collection and the ipfs upload failed for some reasons you can first upload the images to an IPFS provider.

After you received the CID of the image location following steps are required: 

1. update your config file `overwriteImageCID` option with your CID.
2. run `nft-art-maker updateImageCID` which will re-generate your metadata files with the image CID.
3. upload your metadata json folder to your preferred location. 

### Try the example

1. Install the `nft-art-maker` globally: `npm install -g nft-art-maker`
2. Go to the `example` directory and run `nft-art-maker generate`
3. Check the output directory which will be created.
4. Test other commands.

### Development

1. clone the repository
2. run `npm install`
3. link the package locally `npm link`
4. rebuild the project with every change `npm run build`
5. you can run tests by `npm run test` (more integration than unit, you would need to provide the nft.storage api key in the `tests/.nftartmakerrc`)

#### License

MIT

#### Contact

- [Twitter](https://twitter.com/JulianCwirko)
- [WWW](https://www.julian.io)
- [Elven Tools - Elrond blockchain NFT launches - Open Source tools](https://www.elven.tools)


CHANGELOG.md
### [6.0.1](https://github.com/juliancwirko/nft-art-maker/releases/tag/v6.0.1) (2023-08-05)
- update canvas lib to be compatible with Node 18

### [6.0.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v6.0.0) (2022-08-07)
- update dependencies
- changed the build process, to be able to use the newest versions of dependencies. This is why it is a major version

### [5.4.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.4.0) (2022-05-22)
- `nft-art-maker updateImageCID`possibility to update the CID when you don't want to use the `nft-art-maker pack` and `nft-art-maker upload` commands. It is an additional way of doing the same operations. Remember to provide the CID in the configuration file using the `overwriteImageCID` key.
- the `edition` field is now optional
- thanks to @dohomi for your awesome contribution

### [5.3.2](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.3.2) (2022-05-21)
- Improvements for upload functionality: It sometimes hangs with massive files, a couple of GB. These improvements should fix it, but we should monitor that further because it depends on the network and IPFS responsiveness at the moment of sending.
- Added a simple progress indicator.

### [5.3.1](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.3.1) (2022-05-10)
- possibility to configure the metadata files extension, by default `.json`, it can also be set to empty string, so no extension. Thanks to @dohomi.

### [5.3.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.3.0) (2022-05-10)
- nft.storage ipfs car upload using your API key
- progress logging
- some basic tests
- thanks to @dohomi for your awesome contribution

### [5.2.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.2.0) (2022-05-08)
- posibility to define the metadata structure

### [5.1.1](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.1.1) (2022-04-06)
- npm audit fix

### [5.1.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.1.0) (2022-02-20)
- more fixes regarding how it works
- Additional command: `check` will show how many unique assets are generated. Sometimes the names of files can be misleading when there are not enough layers to achieve the required amount of assets. This probably needs some rewrites in the future.
- better output messages which contain the actual number of generated assets

### [5.0.1](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.0.1) (2022-02-20)
- fix the bug when there are no rarity values in the file names

### [5.0.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v5.0.0) (2022-01-05)
- Changes in the metadata file structure, the attributes field is now in the root level
- `shuffleLayerConfigurations` is now enabled by default - in most cases, it is desirable, but you can still disable it in the config file
- it should also now work with Node 17 too

### [4.0.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v4.0.0) (2021-12-26)
- The previous assumption about unique CIDs for every file is not well suited for a proper smart contract with a randomized minting process. It would be hard to keep tracking all the CIDs there. It is much simpler and cheaper to have file numbers for all of them and base CID for the directory. `nft-art-maker pack` will now create two .car files with images and metadata json files. The base CID for images will be replaced in all metadata files after packing them into .car. Then the metadata json files CID will be saved into the main big metadata.json file. You can then use both when minting. (Unique CIDs will still work with version 3.0.0, so feel free to use it if needed.)
- added new fields for the image in the metadata output

### [3.0.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v3.0.0) (2021-11-16)
- schema for metadata.json file changes - it is now more standardized
- possibility to pack files using ipfs-car and generate metadata JSON file for each of them with updated CIDs to files. Metadata files will also be packed, and there will be output with the CIDs list. It could be useful when minting
- ~~removed the option to replace the base file URI. It was ok, but a different URI for each file with a different CID is a better solution, and this is now possible using the `nft-art-maker pack` command~~ --> Invalidated in v4.0.0 (for more details see description for v4)
- these are breaking changes, so the major version is now 3. You can still use the older versions.
- minimum version of Node is now **14.14.0**

### [2.2.2](https://github.com/juliancwirko/nft-art-maker/releases/tag/v2.2.2) (2021-11-08)
- bugfix: version command, now it is possible to check the version without configuration

### [2.2.1](https://github.com/juliancwirko/nft-art-maker/releases/tag/v2.2.1) (2021-11-04)
- bugfix: there was a problem with `shuffleLayerConfigurations` which randomized the entries in the metadata.json file but not the actual order of images. From now the order in the metadata.json is by edition, but then the images are correctly randomly ordered

### [2.2.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v2.2.0) (2021-10-29)
- possibility to update the image path base in metadata.json - useful when you need to update your paths after uploading, for example, when uploading a CAR file using IPFS, but it can be anything.

### [2.1.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v2.1.0) (2021-10-28)
- bring back `shuffleLayerConfigurations` it has a valid use case even if we want only basic functionality
- fix for preview generation

### [2.0.0](https://github.com/juliancwirko/nft-art-maker/releases/tag/v2.0.0) (2021-10-23)
- changes in the output JSON file
- replace sha1 with sha256, which is now also used for provenance hash

src/config.ts
import { cosmiconfigSync } from 'cosmiconfig';
import { cwd } from 'process';

const explorerSync = cosmiconfigSync('nftartmaker');
const customConfig = explorerSync.search(cwd());

// You can switch between default png files and encoded SVG source
const svgBase64DataOnly =
  Boolean(customConfig?.config?.svgBase64DataOnly) || false;

// For more info check: https://github.com/juliancwirko/nft-art-maker#layers-configuration
const layerConfigurations = customConfig?.config?.layerConfigurations;

// Width and height of the image
const format = customConfig?.config?.format || {
  width: 20,
  height: 20,
};

const defaultMetadataSchemaMapper = {
  name: 'name',
  description: 'description',
  edition: 'edition',
  attributes: 'attributes',
  base64SvgDataUri: 'base64SvgDataUri',
  'image.href': 'image.href',
  'image.hash': 'image.hash',
  'image.ipfsUri': 'image.ipfsUri',
  'image.ipfsCid': 'image.ipfsCid',
  'image.fileName': 'image.fileName',
};

const metadataSchemaMapper =
  customConfig?.config?.metadataSchemaMapper || defaultMetadataSchemaMapper;

// Rarity delimiter for a file name
const rarityDelimiter = customConfig?.config?.rarityDelimiter || '#';

// Amount of tries to be able to achieve required size of the collection
const uniqueDnaTorrance = customConfig?.config?.uniqueDnaTorrance || 10000;

// Collection name
const description =
  customConfig?.config?.description || 'Your collection name here';

const nftStorageApiToken =
  customConfig?.config?.nftStorageApiToken || 'Your nft.storage API token';

// Default names for the directories and files
const layersDirName = customConfig?.config?.layersDirName || 'layers';
const outputDirName = customConfig?.config?.outputDirName || 'output';
const outputJsonDirName = customConfig?.config?.outputJsonDirName || 'json';
const outputImagesDirName =
  customConfig?.config?.outputImagesDirName || 'images';
const outputPackedDirName =
  customConfig?.config?.outputPackedDirName || 'packed';
const outputJsonFileName =
  customConfig?.config?.outputJsonFileName || 'metadata.json';
const outputImagesCarFileName =
  customConfig?.config?.outputCarFileName || 'images.car';
const outputMetadataCarFileName =
  customConfig?.config?.outputCarFileName || 'metadata.car';
const outputMetadataFileExtension =
  typeof customConfig?.config?.outputMetadataFileExtension === 'string'
    ? customConfig?.config?.outputMetadataFileExtension
    : '.json';

const overwriteImageCID = customConfig?.config?.overwriteImageCID;

const editionNameFormat = customConfig?.config?.editionNameFormat || '#';

const shuffleLayerConfigurations =
  customConfig?.config?.shuffleLayerConfigurations || true;

// Parameters for the preview file
const preview = customConfig?.config?.preview || {
  thumbPerRow: 20,
  thumbWidth: 60,
  imageRatio: format.width / format.height,
  imageName: 'preview.png',
};

// Tags for your collection as string. Format: 'tag1,tag2,tag3'.
const tags = customConfig?.config?.tags || '';

const config = {
  format,
  description,
  uniqueDnaTorrance,
  layerConfigurations,
  rarityDelimiter,
  preview,
  svgBase64DataOnly,
  layersDirName,
  outputDirName,
  outputJsonDirName,
  outputImagesDirName,
  outputJsonFileName,
  outputPackedDirName,
  outputImagesCarFileName,
  outputMetadataCarFileName,
  editionNameFormat,
  shuffleLayerConfigurations,
  tags,
  metadataSchemaMapper,
  defaultMetadataSchemaMapper,
  nftStorageApiToken,
  outputMetadataFileExtension,
  overwriteImageCID,
};

export default config;
